x/gamm/keeper/migrate.go

```diff
@@ -0,0 +1,142 @@
+package keeper
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/osmosis-labs/osmosis/osmoutils"
+	"github.com/osmosis-labs/osmosis/v14/x/gamm/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// GetMigrationInfo returns the balancer to gamm pool migration info from the store
+func (k Keeper) GetMigrationInfo(ctx sdk.Context) types.MigrationRecords {
+	store := ctx.KVStore(k.storeKey)
+	migrationInfo := types.MigrationRecords{}
+	osmoutils.MustGet(store, types.KeyMigrationInfo, &migrationInfo)
+	return migrationInfo
+}
+
+// SetMigrationInfo sets the balancer to gamm pool migration info to the store
+func (k Keeper) SetMigrationInfo(ctx sdk.Context, migrationInfo types.MigrationRecords) {
+	store := ctx.KVStore(k.storeKey)
+	osmoutils.MustSet(store, types.KeyMigrationInfo, &migrationInfo)
+}
+
+// validateRecords validates a list of BalancerToConcentratedPoolLink records to ensure that:
+// 1) there are no duplicates,
+// 2) the records are in sorted order.
+// 3) both the balancer and gamm pool IDs are valid
+func (k Keeper) validateRecords(ctx sdk.Context, records ...types.BalancerToConcentratedPoolLink) error {
+	lastBalancerPoolID := uint64(0)
+	balancerIdFlags := make(map[uint64]bool)
+	clIdFlags := make(map[uint64]bool)
+
+	for _, record := range records {
+		// If the balancer ID has already been seen, we have a duplicate
+		if balancerIdFlags[record.BalancerPoolId] {
+			return fmt.Errorf(
+				"Balancer pool ID #%d has duplications.",
+				record.BalancerPoolId,
+			)
+		}
+
+		// If the concentrated ID has already been seen, we have a duplicate
+		if clIdFlags[record.ClPoolId] {
+			return fmt.Errorf(
+				"Concentrated pool ID #%d has duplications.",
+				record.ClPoolId,
+			)
+		}
+
+		// Ensure records are sorted from lowest to highest balancer pool ID
+		if record.BalancerPoolId < lastBalancerPoolID {
+			return fmt.Errorf(
+				"Balancer pool ID #%d came after Balancer pool ID #%d.",
+				record.BalancerPoolId, lastBalancerPoolID,
+			)
+		}
+
+		// Ensure the provided balancerPoolId exists and that it is of type balancer
+		poolType, err := k.GetPoolType(ctx, record.BalancerPoolId)
+		if err != nil {
+			return err
+		}
+		if poolType.String() != "Balancer" {
+			return fmt.Errorf("Balancer pool ID #%d is not of type balancer", record.BalancerPoolId)
+		}
+
+		// Ensure the concentrated pool exists. If clPoolID is 0, this signals a removal, so we skip this check.
+		// TODO: Get GetPoolType to work for cl pools from gamm
+		if record.ClPoolId != 0 {
+			_, err = k.poolManager.GetPoolModule(ctx, record.ClPoolId)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
Note to reviewer: We currently have no way that I am aware of to check if a pool is of type Concentrated from the gamm module (like we do in the case above this). Please let me know if I am incorrect.

Reply:
We should have a `GetPoolType` defined on the `poolmanager.PoolI`.

Then, we should get the pool from state and get its type

Reply:
We dont have GetPoolType defined for cl pools, I think we will have to add this

Reply:
For cl pools its only defined in CreatePoolMsg https://github.com/osmosis-labs/osmosis/blob/34626d8dcac8364814486c44e4e1e3341cbe1273/x/concentrated-liquidity/model/msgs.go#L107-L109

Reply:
I added GetPoolType to cl keeper and added respective tests here!

https://github.com/osmosis-labs/osmosis/pull/4071/commits/df4696eccd2802d17524f07b8d99091f0b7a4dd2